### PR TITLE
Parse sourcetext endpoint

### LINKED ISSUE #16876: A more flexible parse endpoint in the TransparentCompiler
**Is your feature request related to a problem? Please describe.**

Currently, the new transparent compiler endpoint for parsing a file does not accommodate all the scenarios we're interested in.

At present, we have:

```fsharp
    /// <summary>
    /// Parses the source code of a file and caches the results. Returns an AST that can be traversed for various features.
    /// </summary>
    ///
    /// <param name="fileName">The file path. The file name serves as a module name for implicit top-level modules (e.g., in scripts).</param>
    /// <param name="sourceText">The source code to be parsed.</param>
    /// <param name="options">Parsing options for the project or script.</param>
    /// <param name="cache">Stores the parse in a size-limited cache associated with the FSharpChecker. Default: true.</param>
    /// <param name="userOpName">An optional string for tracing compiler operations related to this request.</param>
    member ParseFile:
        fileName: string * sourceText: ISourceText * options: FSharpParsingOptions * ?cache: bool * ?userOpName: string ->
            Async<FSharpParseFileResults>

    [<Experimental("This FCS API is experimental and subject to change.")>]
    member ParseFile:
        fileName: string * projectSnapshot: FSharpProjectSnapshot * ?userOpName: string -> Async<FSharpParseFileResults>

    /// <summary>
    /// Parses the source code of a file. Returns an AST for feature traversal.
    /// </summary>
    ///
    /// <param name="fileName">The file path. The file name is also used as a module name for implicit top-level modules (e.g., in scripts).</param>
    /// <param name="source">The source code to be parsed.</param>
    /// <param name="options">Parsing options for the project or script.</param>
    /// <param name="cache">Stores the parse in a size-limited cache associated with the FSharpChecker. Default: true.</param>
    /// <param name="userOpName">An optional string for tracing compiler operations related to this request.</param>
    [<Obsolete("Please use checker.ParseFile instead. This requires passing FSharpParsingOptions instead of FSharpProjectOptions. If needed, generate FSharpParsingOptions from FSharpProjectOptions by calling checker.GetParsingOptionsFromProjectOptions(options).")>]
    member ParseFileInProject:
        fileName: string * source: string * options: FSharpProjectOptions * ?cache: bool * ?userOpName: string ->
            Async<FSharpParseFileResults>
```

The ParseFile endpoint accepting fileName and ISourceText is only available in the old BackgroundCompiler. To phase out the BackgroundCompiler, this functionality needs to be implemented in the new TransparentCompiler.

**Describe the solution you'd like**

From Rider's perspective, parsing typically occurs in three scenarios:

- Parsing the current file in a project when opening a file.
- Parsing sample source code in a sandbox for code actions/quick fixes before inserting code.
- Parsing the current file and inserting an additional identifier at the cursor position to predict potential syntax tree changes and user input. It's crucial that this parsing is cached separately from the file/project cache to ensure the results are not mistakenly reused for ParseAndCheckResults, as they would not be accurate.

Given FileA.fs in ProjectB.fsproj, the following calls should be feasible:

- ParseFileInProject(FileA.fs, snapshotB) (cache result reusable for ParseAndCheck)
- ParseFile(FileA-bis.fs, ISourceTextNew, ParsingOptions) (cache result for FileA with an extra identifier)
- ParseSourceText(ISourceTextNew, isSignature, ParsingOptions) (do not cache)

ParseFile and ParseSourceText could potentially merge into a single API, with a boolean parameter to indicate caching preference.

//cc @auduchinok (Please correct me if I misunderstood anything).
//cc @0101 

**Describe alternatives you've considered**

I don't think we can do all of this purely with the TransparentCompiler. Or at the very least, it would today invoke the background compiler.

### PR DESCRIPTION
**Pull request by nojaf**
## Description

Fixes #16876

I've tweaked the `TransparentCompiler` to work better with the [ParseFile](https://fsharp.github.io/fsharp-compiler-docs/reference/fsharp-compiler-codeanalysis-fsharpchecker.html#ParseFile) function, aiming to make things smoother for folks using Rider.

Before this change, `TransparentCompiler` was stuck using an older method, the BackgroundCompiler, for `ParseFile`. Now, we've moved to a different strategy. We're either using a new, smaller cache for storing results or not caching at all. This is pretty handy, especially when you're whipping up ASTs for code generation.

I decided to go with a brand-new cache for this because it's meant to be used without messing with the project's own parsing and caching setup. So, it's all about keeping things neat and tidy, without stepping on the toes of the project-specific file caching.

## Checklist

- [ ] Test cases added
- [ ] Performance benchmarks added in case of performance changes
- [ ] Release notes entry updated:
    > Please make sure to add an entry with short succinct description of the change as well as link to this pull request to the respective release notes file, if applicable.
    >
    > Release notes files:
    > - If anything under `src/Compiler` has been changed, please make sure to make an entry in `docs/release-notes/.FSharp.Compiler.Service/<version>.md`, where `<version>` is usually "highest" one, e.g. `42.8.200`
    > - If language feature was added (i.e. `LanguageFeatures.fsi` was changed), please add it to `docs/releae-notes/.Language/preview.md`
    > - If a change to `FSharp.Core` was made, please make sure to edit `docs/release-notes/.FSharp.Core/<version>.md` where version is "highest" one, e.g. `8.0.200`.

    > Information about the release notes entries format can be found in the [documentation](https://fsharp.github.io/fsharp-compiler-docs/release-notes/About.html).
    > Example:
    > * More inlines for Result module. ([PR #16106](https://github.com/dotnet/fsharp/pull/16106))
    > * Correctly handle assembly imports with public key token of 0 length. ([Issue #16359](https://github.com/dotnet/fsharp/issues/16359), [PR #16363](https://github.com/dotnet/fsharp/pull/16363))
    > *`while!` ([Language suggestion #1038](https://github.com/fsharp/fslang-suggestions/issues/1038), [PR #14238](https://github.com/dotnet/fsharp/pull/14238))

    > **If you believe that release notes are not necessary for this PR, please add `NO_RELEASE_NOTES` label to the pull request.**

### CODE REVIEW

File: src/Compiler/Service/TransparentCompiler.fs [hunk 1]
```diff
@@ -267,6 +267,9 @@ type internal CompilerCaches(sizeFactor: int) =
 
     member val ParseFile = AsyncMemoize(keepStrongly = 50 * sf, keepWeakly = 20 * sf, name = "ParseFile")
 
+    member val ParseFileWithoutProject =
+        AsyncMemoize<string, string, FSharpParseFileResults>(keepStrongly = 10 * sf, keepWeakly = 5 * sf, name = "ParseFileWithoutProject")
```
Comment (auduchinok) [inline src/Compiler/Service/TransparentCompiler.fs:271]: We'd be fine with even a smaller cache in Rider. Maybe even keeping the last file only would already be enough 😅
Maybe make it 5 or something less?

**Review by auduchinok — APPROVED** [review]
Thanks @nojaf!

File: src/Compiler/Service/TransparentCompiler.fs (hunk 1 above)
Comment (nojaf) [reply src/Compiler/Service/TransparentCompiler.fs:271]: Yes, I'm assuming this would really only be used for the single file you are editing right?

Comment (auduchinok) [reply src/Compiler/Service/TransparentCompiler.fs]: Yes.

Comment (nojaf) [reply src/Compiler/Service/TransparentCompiler.fs]: Gotcha, made it smaller.

**Review by psfinaki — APPROVED** [review]

**Pull request merged**